### ISSUE TITLE
Regex-compare number values

### ISSUE THREAD

**Issue by sdaschner** [lukas-krecan/JsonUnit#40]
First of all, thanks for this great tool!

It's currently not possible to compare numbers with regex values, like this -- as the types don't match:

```
assertThat("{\"test\":12}", jsonEquals("{\"test\":\"${json-unit.regex}[0-9]+\"}"));
```

Is it possible to enhance the library to give a more flexible comparison while testing one big JSON at once? IMO it is very productive to just having large JSON strings (e.g. in a file) with the `${json-unit.*}` enhancement for everything which is not fix. That's why I like the idea of these 'variables' very much.

Maybe one of these proposals (which would change the compared type then) -- you certainly have a better view which would fit in the current overall picture:

```
${json-unit.number}[0-9]+
${json-unit.regex.number}[0-9]+
${json-unit.number}${json-unit.regex}[0-9]+
```

WDYT?

**Comment by sdaschner** [comment]
Oh, and I forgot to mention: If you accept PR (and of course think that this is a helpful extension), I'll be happy to contribute and implement this feature.

**Comment by lukas-krecan** [comment]
Thanks for the feedback. It's possible to do, I just have to think about the best way. I just do not want to end-up with a new scripting language.

In the meantime you can use `${json-unit.ignore}` in combination with Hamcrest or Fluent matchers

**Comment by sdaschner** [comment]
Fully agree. A simple and productive way would be nice, just for (ideally) all JSON types.

Yep, I'm doing that. Let me know if you need any help (deciding or implementing).

**Comment by lukas-krecan** [comment]
Here are my thoughts (a.k.a. incoherent brain-dump):

1) Changing regexp implementation so it works also on numbers is quite easy, the question is if it makes sense (do we want it to work on booleans as well?, what about null?)
2) Is the change backward compatible? Most likely yes.
3) Is it more surprising that it works on numbers or that it does not?
4) What's the usecase to use regexp on numbers? It feels a bit silly to use regexp on numbers, even though it might be quite powerfull (imagine regexp checking if the number is odd or even, what a beauty)

Basically I hesitate between two options. 
A) Add support for numbers, booleans and null to ${json-unit.regex} as you have proposed. (I still am not sure about null)
B) Add something like ${json-unit.any-number} or ${json-unit.any-boolean}

Would option B) be good enough for you?

**Comment by lukas-krecan** [comment]
I had some early-morning inspiration and I came up with something like this
```
assertThatJson("{\"test\":{\"a\":1, \"b\":\"value\"}}")
        .withInlineMatcher("isLessThan10", lessThanOrEqualTo(valueOf(10))) // Hamcrest matcher
        .isEqualTo("{test:{a:'${json-unit.inline-matcher:isLessThan10}', b:'value'}}");
```

What do you think?

**Comment by sdaschner** [comment]
Regarding your thoughts:
1) Yes, the best would be to have something like this for all types
2) Yes, by introducing new 'variables', it is
3+4) That's a good point with regex, that's why I would favor your option B), more 'variables' for several use cases / types

I find the option in your last comment a little bit too complex for simple use cases, personally.
+1 for the `${json-unit.any-*}`

**Comment by lukas-krecan-lt** [comment]
Released as 1.13.0, enjoy

**Comment by sdaschner** [comment]
Works as expected, thanks a lot!
